libvips/resample/thumbnail.c

```diff
@@ -760,16 +785,16 @@ vips_thumbnail_build( VipsObject *object )
 	if( vips_image_hasalpha( in ) && 
 	 	hshrink != 1.0 &&
 		vshrink != 1.0  ) { 
-		g_info( "premultiplying alpha" ); 
-		if( vips_premultiply( in, &t[3], NULL ) ) 
-			return( -1 );
-		have_premultiplied = TRUE;
-
 		/* vips_premultiply() makes a float image. When we
 		 * vips_unpremultiply() below, we need to cast back to the
-		 * pre-premultiply format.
+		 * pre-premultiplied format.
 		 */
+		g_info( "premultiplying alpha" ); 
+		have_premultiplied = TRUE;
```

Comment:
Not sure if this is neater, but instead of keeping a separate `have_premultiplied` flag, you could also set `unpremultiplied_format` to `VIPS_FORMAT_NOTSET` and check for `unpremultiplied_format != VIPS_FORMAT_NOTSET` instead.

Reply:
Good idea!